### DISCUSSION TITLE
How to define custom token types?

### DISCUSSION THREAD

**Discussion by dbitouze** [pygments/pygments#2566] [Q&A]
AFAICS, `TexLexer` seems to need some improvements. `Pygments`' built-in tokens are not appropriate for the `(La)TeX` language, so I'm trying to declare custom token types. I didn't see anything about this in the documentation.

- Fortunately, there's issue #1840.
- Unfortunately, I didn't understand how @jeanas managed to create custom tokens such as [`Token.ChordModifier`](https://github.com/pygments/pygments/blob/f98faa6d9b97f1c58df0c9b5c609fc21db193fa9/pygments/lexers/lilypond.py#L167C60-L167C79).

My attempt, inspired by issue #1986, is as follows:

```python
import pygments
from pygments.lexer import RegexLexer
from pygments.token import *
from pygments.style import Style
from pygments.formatters.terminal256 import Terminal256Formatter

class TestLexer(RegexLexer):
    tokens = {
        'root': [
            (r'A', Name),
            (r'B', Name.Function.CUSTOM),
            (r'C', MyCustomToken),
        ]
    }

class MyStyle(Style):
    styles = {
        Name: "#FA9BFA",
        Name.Function.CUSTOM: "#FA9BFA",
        MyCustomToken: "#FA9BFA"
    }

print(pygments.highlight(
    code="ABC",
    lexer=TestLexer(),
    formatter=Terminal256Formatter(style=MyStyle),
).strip())
```

but it fails with the error:

```
Traceback (most recent call last):
  File "/home/bitouze/Public/Repositories/pygments-acetexlexer/lexer/custom-tokens.py", line 7, in <module>
    class TestLexer(RegexLexer):
  File "/home/bitouze/Public/Repositories/pygments-acetexlexer/lexer/custom-tokens.py", line 12, in TestLexer
    (r'C', MyCustomToken),
NameError: name 'MyCustomToken' is not defined
```

What am I missing?

**Comment by jeanas** [comment]
Just use `Token.MyCustomToken`, or define `MyCustomToken = Token.MyCustomToken` and then use `MyCustomToken`.

See https://pygments.org/docs/tokens/: “You can create a new token type by accessing an attribute of Token whose name starts with an uppercase letter”.

**Comment by birkenfeld** [comment]
Making custom token types is easy. The problem is in that they then won't be styled in any style.

So don't worry about the builtin types not perfectly matching semantically, it's still preferable to use them whenever possible in some way.

**Comment by dbitouze** [comment]
> Just use `Token.MyCustomToken`, or define `MyCustomToken = Token.MyCustomToken` and then use `MyCustomToken`.
> 
> See https://pygments.org/docs/tokens/: “You can create a new token type by accessing an attribute of Token whose name starts with an uppercase letter”.

That's what I tried but it didn't work and now I understand: in my real use case, I had to define `MyCustomToken` both in the lexer and in the style in order to make it work.

Thanks!

**Comment by dbitouze** [comment]
> Making custom token types is easy. The problem is in that they then won't be styled in any style.

@birkenfeld Indeed.
 
> So don't worry about the builtin types not perfectly matching semantically, it's still preferable to use them whenever possible in some way.

How ugly! :wink: 

Currently, the `TexLexer` involves the following builtin tokens:

- `Comment`
- `Keyword`
- `Name.Attribute`
- `Name.Builtin`
- `Name.Variable`
- `Number`
- `Operator`
- `String`
- `String.Backtick`
- `Text`

Wouldn't it be possible for `Pygments` to declare new builtin tokens dedicated to `(La)TeX` and, in each of the existing styles files, to declare them to be aliases to the ones above. Hence, for `(La)TeX` codes:

- existing styles will continue to work,
- future styles could take in account these builtin tokens dedicated to `(La)TeX` to be more appropriate and possibly more accurate.

**Comment by birkenfeld** [comment]
You may call it ugly, but I'd call token types used for only one lexer ugly.

It would not only create work now to adapt all existing styles, but also for each future style, with no clear benefit.

**Comment by jeanas** [comment]
IIRC, for LilyPond, I didn't manage to get good results by using standard token types, thus the special types and the special-purpose style.

If possible, sticking with standard tokens is indeed preferable.

If useful, it's always possible to use subtokens like (say) `Token.Name.Class.Environment`, which would get styled as `Token.Name.Class` if the style doesn't specifically define `Token.Name.Class.Environment`.
